File: glimpse/services/unsplash_service.py
"""Unsplash API service for fetching random images"""

import requests
from typing import Dict


class UnsplashService:
    """Service for interacting with Unsplash API"""

    def __init__(self, access_key: str):
        if not access_key:
            raise ValueError('Unsplash access key is required')
        self.access_key = access_key
        self.base_url = 'https://api.unsplash.com'
        self.name = 'unsplash'

    def fetch_random_image(self) -> Dict[str, str]:
        """
        Fetch a random image from Unsplash

        Returns:
            Dictionary with url, description, and sourceUrl keys

        Raises:
            Exception: If the API request fails
        """
        try:
            response = requests.get(
                f'{self.base_url}/photos/random',
                headers={'Authorization': f'Client-ID {self.access_key}'},
                params={'orientation': 'landscape'},
                timeout=30
            )
            response.raise_for_status()

            photo = response.json()

            return {
                'url': photo['urls']['regular'],
                'description': photo.get('description') or photo.get('alt_description') or 'No description available',
                'sourceUrl': photo['links']['html']
            }
        except requests.exceptions.RequestException as e:
            raise Exception(f'Failed to fetch image from Unsplash: {str(e)}')

    def test_connection(self) -> bool:
        """Test the Unsplash API connection"""
        try:
            response = requests.get(
                f'{self.base_url}/photos/random',
                headers={'Authorization': f'Client-ID {self.access_key}'},
                timeout=10
            )
            response.raise_for_status()
            return True
        except Exception:
            return False

File: src/services/unsplash-service.ts
import axios from 'axios';
import { ImageSource, UnsplashPhoto } from '../types';
import { logger } from '../utils/logger';

export class UnsplashService implements ImageSource {
  name = 'unsplash';
  private accessKey: string;
  private baseUrl = 'https://api.unsplash.com';

  constructor(accessKey: string) {
    if (!accessKey) {
      throw new Error('Unsplash access key is required');
    }
    this.accessKey = accessKey;
  }

  async fetchRandomImage(): Promise<{
    url: string;
    description: string;
    sourceUrl: string;
  }> {
    try {
      logger.debug('Fetching random image from Unsplash...');

      const response = await axios.get<UnsplashPhoto>(`${this.baseUrl}/photos/random`, {
        headers: {
          Authorization: `Client-ID ${this.accessKey}`,
        },
        params: {
          orientation: 'landscape',
        },
      });

      const photo = response.data;

      return {
        url: photo.urls.regular,
        description:
          photo.description || photo.alt_description || 'No description available',
        sourceUrl: photo.links.html,
      };
    } catch (error) {
      if (axios.isAxiosError(error)) {
        logger.error('Unsplash API error:', error.response?.data || error.message);
        throw new Error(`Failed to fetch image from Unsplash: ${error.message}`);
      }
      throw error;
    }
  }

  /**
   * Test the Unsplash API connection
   */
  async testConnection(): Promise<boolean> {
    try {
      await axios.get(`${this.baseUrl}/photos/random`, {
        headers: {
          Authorization: `Client-ID ${this.accessKey}`,
        },
      });
      return true;
    } catch (error) {
      return false;
    }
  }
}
